frontend/apps/reader/modules/readerhighlight.lua

```diff
@@ -297,6 +306,47 @@ end
 
 function ReaderHighlight:onReaderReady()
     self:setupTouchZones()
+    self:migrateHighlights()
+end
+
+
+function ReaderHighlight:migrateHighlights()
+    if self.ui.doc_settings:hasNot("highlight_merged") then
+        for page, highlights in pairs(self.view.highlight.saved) do
+            for _, item in ipairs(highlights) do
+                -- This is to switch notes and text.
+                local highlight = bookmarkToLocalHighlight(item)
+                if not highlight.chapter then
+                    local pg_or_xp = self.ui.paging and page or highlight.pos0
+                    local chapter_name = self.ui.toc:getTocTitleByPage(pg_or_xp)
+                    highlight.chapter = chapter_name
+                end
+                if not highlight.drawer then
+                    highlight.drawer = self.view.highlight.saved_drawer
+                end
+                highlight.page_num = page
+                if self.ui.paging then
+                    highlight.page = page
+                end
+                self.ui.bookmark:patchBookmark({
+                    datetime = highlight.datetime,
+                    page = highlight.page,
+                    highlight
+                });
+            end
+        end
+        for i, bookmark in ipairs(self.ui.bookmark.bookmarks) do
+            if bookmark.highlighted then
+                if not bookmark.drawer then
+                    self.ui.bookmark.bookmarks[i].drawer = self.view.highlight.saved_drawer
+                end
+                if not bookmark.chapter and bookmark.pos0 or bookmark.page then
```

Comment:
Is it correct? `and` prevails `or`.
In any case, `bookmark.page` always exists in bookmarks.

Reply:
If it does always exists then I'm removing this part.